internal/appliance/gitserver.go

```diff
@@ -0,0 +1,135 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
+
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/container"
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/pod"
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/pvc"
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/service"
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/serviceaccount"
+	"github.com/sourcegraph/sourcegraph/internal/k8s/resource/statefulset"
+)
+
+func (r *Reconciler) reconcileGitServer(ctx context.Context, sg *Sourcegraph, owner client.Object) error {
+	if err := r.reconcileGitServerStatefulSet(ctx, sg, owner); err != nil {
+		return err
+	}
+	if err := r.reconcileGitServerService(ctx, sg, owner); err != nil {
+		return err
+	}
+	if err := r.reconcileGitServerServiceAccount(ctx, sg, owner); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileGitServerStatefulSet(ctx context.Context, sg *Sourcegraph, owner client.Object) error {
+	cfg := sg.Spec.GitServer
+	name := "gitserver"
+
+	ctr := container.NewContainer(name, cfg, corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceCPU:    resource.MustParse("4"),
+			corev1.ResourceMemory: resource.MustParse("8Gi"),
+		},
+		Limits: corev1.ResourceList{
+			corev1.ResourceCPU:    resource.MustParse("4"),
+			corev1.ResourceMemory: resource.MustParse("8Gi"),
+		},
+	})
+
+	storageSize, err := resource.ParseQuantity(cfg.StorageSize)
+	if err != nil {
+		return errors.Wrap(err, "parsing storage size")
+	}
+
+	// TODO: https://github.com/sourcegraph/sourcegraph/issues/62076
+	ctr.Image = "index.docker.io/sourcegraph/gitserver:5.3.2@sha256:6c6042cf3e5f3f16de9b82e3d4ab1647f8bb924cd315245bd7a3162f5489e8c4"
+
+	ctr.Env = []corev1.EnvVar{
+		container.NewEnvVarSecretKeyRef("REDIS_CACHE_ENDPOINT", "redis-cache", "endpoint"),
+		container.NewEnvVarSecretKeyRef("REDIS_STORE_ENDPOINT", "redis-store", "endpoint"),
+
+		// OTEL_AGENT_HOST must be defined before OTEL_EXPORTER_OTLP_ENDPOINT to substitute the node IP on which the DaemonSet pod instance runs in the latter variable
+		container.NewEnvVarFieldRef("OTEL_AGENT_HOST", "status.hostIP"),
+		{Name: "OTEL_EXPORTER_OTLP_ENDPOINT", Value: "http://$(OTEL_AGENT_HOST):4317"},
+	}
+
+	ctr.Ports = []corev1.ContainerPort{
+		{Name: "rpc", ContainerPort: 3178},
+	}
+
+	ctr.LivenessProbe = &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			TCPSocket: &corev1.TCPSocketAction{
+				Port: intstr.FromString("rpc"),
+			},
+		},
+		TimeoutSeconds:      5,
+		InitialDelaySeconds: 5,
+	}
+
+	ctr.VolumeMounts = []corev1.VolumeMount{
+		{Name: "tmpdir", MountPath: "/tmp"},
+		{Name: "repos", MountPath: "/data/repos"},
+	}
+
+	podVolumes := []corev1.Volume{
+		pod.NewVolumeEmptyDir("repos"),
```

Comment:
Back in https://github.com/sourcegraph/sourcegraph/pull/62270#issuecomment-2085020236, I commented that we didn't need an emptyDir volume for something because it would be shadowed by the mounted PVC. I was only half-correct: while the emptyDir part would have been ignored and shadowed by the PVC, we do need a named, typeless volume here so that the statefulset knows which named PVC to mount here:

```suggestion
		{Name: "repos"},
```

Helm equivalent: https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/templates/gitserver/gitserver.StatefulSet.yaml#L101

This does need fixing in symbols too, my bad 🤦

Reply:
Ahh okay, I was wondering a bit about this as well and went back and forth between two a few times 😄